src/contracts/facilitators/gsm/GsmL2.sol

```diff
@@ -0,0 +1,601 @@
+    uint256 fee = withFee ? IGsmFeeStrategy(_feeStrategy).getBuyFee(grossAmount) : 0;
+    uint256 ghoSold = grossAmount + fee;
+    uint256 finalGrossAmount = withFee
+      ? IGsmFeeStrategy(_feeStrategy).getGrossAmountFromTotalBought(ghoSold)
+      : ghoSold;
+    // pick the lowest asset amount possible for given GHO amount
+    uint256 finalAssetAmount = IGsmPriceStrategy(PRICE_STRATEGY).getGhoPriceInAsset(
+      finalGrossAmount,
+      false
+    );
+    uint256 finalFee = ghoSold - finalGrossAmount;
+    return (finalAssetAmount, finalGrossAmount + finalFee, finalGrossAmount, finalFee);
+  }
+
+  /**
+   * @dev Returns the amount of GHO bought in exchange of a given amount of underlying asset
+   * @param assetAmount The amount of underlying asset to sell
+   * @return The exact amount of asset the user sells
+   * @return The total amount of GHO the user buys (gross amount in GHO minus fee)
+   * @return The gross amount of GHO
+   * @return The fee amount in GHO, applied to the gross amount of GHO
+   */
+  function _calculateGhoAmountForSellAsset(
+    uint256 assetAmount
+  ) internal view returns (uint256, uint256, uint256, uint256) {
+    bool withFee = _feeStrategy != address(0);
+    // pick the lowest GHO amount possible for given asset amount
+    uint256 grossAmount = IGsmPriceStrategy(PRICE_STRATEGY).getAssetPriceInGho(assetAmount, false);
+
+    require(_ghoLiquidity >= grossAmount, Errors.INSUFFICIENT_GHO_LIQ);
+
+    uint256 fee = withFee ? IGsmFeeStrategy(_feeStrategy).getSellFee(grossAmount) : 0;
+    uint256 ghoBought = grossAmount - fee;
+    uint256 finalGrossAmount = withFee
+      ? IGsmFeeStrategy(_feeStrategy).getGrossAmountFromTotalSold(ghoBought)
+      : ghoBought;
+    // pick the highest asset amount possible for given GHO amount
+    uint256 finalAssetAmount = IGsmPriceStrategy(PRICE_STRATEGY).getGhoPriceInAsset(
+      finalGrossAmount,
+      true
+    );
+    uint256 finalFee = finalGrossAmount - ghoBought;
+    return (finalAssetAmount, finalGrossAmount - finalFee, finalGrossAmount, finalFee);
+  }
+
+  /**
+   * @dev Updates Fee Strategy
+   * @param feeStrategy The address of the new Fee Strategy
+   */
+  function _updateFeeStrategy(address feeStrategy) internal {
+    address oldFeeStrategy = _feeStrategy;
+    _feeStrategy = feeStrategy;
+    emit FeeStrategyUpdated(oldFeeStrategy, feeStrategy);
+  }
+
+  /**
+   * @dev Updates Exposure Cap
+   * @param exposureCap The value of the new Exposure Cap
+   */
+  function _updateExposureCap(uint128 exposureCap) internal {
+    uint128 oldExposureCap = _exposureCap;
+    _exposureCap = exposureCap;
+    emit ExposureCapUpdated(oldExposureCap, exposureCap);
+  }
+
+  /**
+   * @dev Updates Liquidity Provider
+   * @param liquidityProvider The address of the liquidty provider for the GSM
+   */
+  function _updateLiquidityProvider(address liquidityProvider) internal {
```

Comment:
does it make sense to use a role for this instead?

Reply:
i was thinking only one facilitator can provide liquidity to the GSM so that's why I went with one address but don't have a strong opinion on it